src/stdlib_64_bit_spookyv2_hashes.fypp

```diff
@@ -0,0 +1,718 @@
+        integer(${k1}$), intent(in) :: key(:)
+        integer(int64), intent(in)  :: seed(2)
+        integer(int64)              :: hash_code(2)
+
+        integer(int64) :: hash2(2)
+
+        hash2(:) = seed
+        call spookyhash_128( transfer( key, 0_int8, &
+                     bytes_${k1}$*size(key, kind=int64) ), hash2 )
+        hash_code = hash2
+
+    end function ${k1}$_spooky_hash
+
+#:endfor
+
+     module function character_spooky_hash( key, seed ) result(hash_code)
+        character(*), intent(in)   :: key
+        integer(int64), intent(in) :: seed(2)
+        integer(int64)             :: hash_code(2)
+
+        integer(int64)    :: hash2(2)
+
+        hash2(:) = seed
+        call spookyhash_128( transfer( key, 0_int8,                    &
+                             bytes_char*len(key, kind=int64) ), hash2 )
+        hash_code = hash2
+
+    end function character_spooky_hash
+
+!
+! short hash ... it could be used on any message,
+! but it's used by Spooky just for short messages.
+!
+     subroutine spookyhash_short( key, hash_inout )
+        integer(int8), intent(in), target :: key(0:)
+        integer(int64), intent(inout)     :: hash_inout(2)
+
+        integer(int64) :: a, b, c, d
+        integer(int64) :: length, p8, remainder
+
+        p8 = 0
+        length = size( key, kind=int64 )
+
+        ! The number of bytes after all the INT256s
+        remainder = iand( length, 31_int64 )
+        a = hash_inout(1)
+        b = hash_inout(2)
+        c = sc_const
+        d = sc_const
+
+        if ( length > 15 ) then
+            block
+                integer(int64) :: bend, step
+                integer(int64) :: buf(0:2*sc_numVars-1)
+                bend = ishft(length, -4) ! The number of complete INT128s
+                buf(0:2*bend-1) = transfer( key(0:16*bend-1), 0_int64, 2*bend )
+                ! Number of Int64's in number of complete INT256s
+                bend = ishft(ishft(length, -5), 2)
+
+                ! handle all complete sets of 32 bytes
+                do step = 0_int64, bend-1, 4
+                    c = c + buf(step)
+                    d = d + buf(step+1)
+                    call shortmix( a, b, c, d )
+                    a = a + buf(step+2)
+                    b = b + buf(step+3)
+                end do
+                ! Completed all INT64s in complete INT256s
+                p8 = p8 + 8*bend ! Number of INT8s in complete INT256s
+
+                ! Handle the case of 16+ remaining bytes.
+                if (remainder >= 16) then
+                    c = c + buf(step)
+                    d = d + buf(step+1)
+                    call shortmix( a, b, c, d )
+                    p8 = p8 + 16
+                    remainder = remainder - 16
+                end if
+
+            end block
+        end if
+
+    ! Handle the last 0..15 bytes, and its length V2
+        d = d + shiftl( length, 56_int64 )
+
+        select case(remainder)
+        case(15)
+            go to 115
```

Comment:
In this situation, is there any natural alternative to the goto's? I do not wish to be dogmatic about it -- but if it can be avoided without much complication, that is good.

Reply:
The corresponding C code was implemented with switches with no breaks. I could implement it with a computed go to but that is not encouraged in the current language. I could also implement it with a sequence of if statements, but that introduces some overhead for unnecessary tests, unless the compiler is very smart about optimization.